Katello/katello
# Fixes #24166 - Add requiredTags to custom repos

This PR allows you to add __OS Versions__ to repositories in custom products.  On RHEL systems, this will automatically disable the repo on any system that doesn't have a matching "provided tag."

### Repo edit screen
![no_restrict](https://user-images.githubusercontent.com/22042343/97585226-47e44900-19cf-11eb-8499-a0cf4e22009d.png)

### New repository screen
![new_repo_norestrict](https://user-images.githubusercontent.com/22042343/97585353-7104d980-19cf-11eb-89fe-fd470f752e9f.png)

## How it works

Consider a custom product with custom repository.  Traditionally these custom repos are always enabled by default.  With this new feature, however, repos can be disabled for certain OS versions via Candlepin's `requiredTags` field.

#### Example with RHEL 7 repo and RHEL 7 host
1. A custom repo is created with "Restrict to OS Version" set to "Red Hat Enterprise Linux 7"
2. This causes the Candlepin product to be updated with `requiredTags` set to `rhel-7`
3. On a RHEL 7 host, there exists a product certificate in `/etc/pki/product` which has "provided tags" `rhel-7,rhel-7-server`
4. Since the host provides all of the required tags, the repo is enabled by default.

#### Example with RHEL 7 repo and RHEL 6 host
1. A custom repo is created with "Restrict to OS Version" set to "Red Hat Enterprise Linux 7"
2. This causes the Candlepin product to be updated with `requiredTags` set to `rhel-7`
3. On a RHEL 6 host, there exists a product certificate in `/etc/pki/product` which has "provided tags" `rhel-6,rhel-6-server`, but not `rhel-7`
4. Since the host does not provide the required tags, the repo is disabled.

#### Example with CentOS 7 host
1. A custom repo is created with "Restrict to OS Version" set to "Red Hat Enterprise Linux 7"
2. This causes the Candlepin product to be updated with `requiredTags` set to `rhel-7`
3. On the CentOS host, no matching product certificate exists
4. Since the host does not provide the required tags, the repo is disabled.

__Note:__ For custom repositories with _no_ OS versions set, the product will be enabled by default regardless of OS version.  This is not a change from previous behavior.

| Provided tags (on host) | Required tags (repo) | Result |
| ------------- | ------------- | ------------- |
| rhel-7, rhel-7-server | rhel-7 | Enabled
| rhel-7, rhel-7-server | (none) | Enabled
| rhel-7, rhel-7-server | rhel-8 | Disabled
| rhel-7, rhel-7-server | rhel-7, rhel-8 | Disabled*
| (none) | rhel-7 | Disabled
| (none) | rhel-7, rhel-8 | Disabled
| (none) | (none) | Enabled

As we can see, required tags are _required_; they live up to their name.  In order for the custom repo to be enabled, a content host must provide __all__ of the required tags.  Therefore, this feature will only be useful for RHEL hosts.

*You may think that adding multiple required tags would allow a repo to be enabled on multiple OS versions.  But this is not how it works.  A repo with both `rhel-7` and `rhel-8` tags would be disabled on all systems, because there are no systems that would list both `rhel-7` and `rhel-8` in 'Provided Tags'.  For this reason, this feature only sets a single required tag and does not allow multiples.

## To test

### Web UI

* Create a custom product and repo
* Set 'Restrict to OS Version'
* Save the repo
* View the repo details and verify 'Restrict to OS Version' value was saved

### Check that Candlepin is updated correctly

```ruby
# rails console
prod_all = Katello::Resources::Candlepin::Product.all('Default_Organization').select { |p| !p['productContent'].nil? }.map { |q| q['productContent'] }.flatten.uniq
prod_all.find { |p| p['content']['name'] == 'My New Repo' }

=> {"content"=>
  {"created"=>"2020-10-12T19:23:43+0000",
   "updated"=>"2020-10-19T16:22:09+0000",
   "uuid"=>"4028face7504ef7e017541ab07e60565",
   "id"=>"1602530623303",
   "type"=>"yum",
   "label"=>"Default_Organization_SomeProduct_asdf",
   "name"=>"asdf",
   "vendor"=>"Custom",
   "contentUrl"=>"/custom/SomeProduct/asdf",
   "requiredTags"=>"rhel-8,rhel-7-server,rhel-7-workstation",
   "gpgUrl"=>nil,
   "modifiedProductIds"=>[],
   "arches"=>nil,
   "requiredProductIds"=>[],
   "metadataExpire"=>1,
   "releaseVer"=>nil},
 "enabled"=>true}
```
Look for the `requiredTags` value.

### Test with a host
_Note:_ The Restrict to OS Version feature only works with RHEL hosts.  However, for testing purposes you can make a CentOS host look like a RHEL host by obtaining the RHEL product certificate in step 2 below.  On actual RHEL hosts, you can skip step 2.

1. Register a host to Katello.
2. Obtain a certificate from https://access.redhat.com/labs/rhpc/  (recommended: Red Hat Enterprise Linux Server ->  Version 7.9 -> x86_64 -> Generate)
3. Ensure your new product certificate is in `/etc/pki/product`

```
% rct cat-cert /etc/pki/product/FOO.pem
Product:
    ID: 69
    Name: Red Hat Enterprise Linux Server
    Version: 7.5
    Arch: x86_64
    Tags: rhel-7,rhel-7-server
    Brand Type: 
    Brand Name:
```

This confirms that the "Provided Tags" on your host are now `rhel-7,rhel-7-server`.

4. Save a custom product & repo with 'Restrict to OS Version' set to 'Red Hat Enterprise Linux 7'.  This will cause the Candlepin product content to have the `rhel-7` tag.
5. Ensure the host is subscribed to that product
6. `subscription-manager refresh` on the host
7.  Find and view the entitlement certificate for the custom product:
```
% rct cat-cert /etc/pki/entitlements/FOO.pem
Content:
	Type: yum
	Name: repo7
	Label: Default_Organization_SomeProduct_repo7
	Vendor: Custom
	URL: /Default_Organization/Library/custom/SomeProduct/repo7
	GPG: 
	Enabled: True
	Expires: 1
	Required Tags: rhel-7
	Arches: ALL
```

The Required Tags should match what was saved in the web UI.
 
Since the 'Required Tags' are a subset of the 'Provided Tags', `subscription-manager repos` should show the repo as enabled:
```
Repo ID:   Default_Organization_SomeProduct_repo7
Repo Name: repo7
Repo URL:  https://centos7-katello-devel.redhatlaptop.example.com/pulp/repos/Default_Organization/Library/custom/SomeProduct/repo7
Enabled:   1
```

Test again with mismatching Required Tags:

```
subscription-manager refresh; subscription-manager repos
```
The repo should be disabled (`Enabled: 0`)
